cmd/kubeadm/app/cmd/phases/join/kubelet.go

```diff
@@ -150,44 +165,57 @@ func runKubeletStartJoinPhase(c workflow.RunData) (returnErr error) {
-	if err := kubeletphase.WriteConfigToDisk(&initCfg.ClusterConfiguration, kubeadmconstants.KubeletRunDirectory); err != nil {
+	if err := kubeletphase.WriteConfigToDisk(&initCfg.ClusterConfiguration, data.KubeletDir()); err != nil {
 		return err
 	}
 
 	// Write env file with flags for the kubelet to use. We only want to
 	// register the joining node with the specified taints if the node
 	// is not a control-plane. The mark-control-plane phase will register the taints otherwise.
 	registerTaintsUsingFlags := cfg.ControlPlane == nil
-	if err := kubeletphase.WriteKubeletDynamicEnvFile(&initCfg.ClusterConfiguration, &initCfg.NodeRegistration, registerTaintsUsingFlags, kubeadmconstants.KubeletRunDirectory); err != nil {
+	if err := kubeletphase.WriteKubeletDynamicEnvFile(&initCfg.ClusterConfiguration, &initCfg.NodeRegistration, registerTaintsUsingFlags, data.KubeletDir()); err != nil {
 		return err
 	}
 
 	// Try to start the kubelet service in case it's inactive
-	fmt.Println("[kubelet-start] Starting the kubelet")
-	kubeletphase.TryStartKubelet()
-
-	// Now the kubelet will perform the TLS Bootstrap, transforming /etc/kubernetes/bootstrap-kubelet.conf to /etc/kubernetes/kubelet.conf
-	// Wait for the kubelet to create the /etc/kubernetes/kubelet.conf kubeconfig file. If this process
-	// times out, display a somewhat user-friendly message.
-	waiter := apiclient.NewKubeWaiter(nil, kubeadmconstants.TLSBootstrapTimeout, os.Stdout)
-	if err := waiter.WaitForKubeletAndFunc(waitForTLSBootstrappedClient); err != nil {
-		fmt.Printf(kubeadmJoinFailMsg, err)
-		return err
-	}
+	if !data.DryRun() {
+		fmt.Println("[kubelet-start] Starting the kubelet")
+		kubeletphase.TryStartKubelet()
 
-	// When we know the /etc/kubernetes/kubelet.conf file is available, get the client
-	client, err := kubeconfigutil.ClientSetFromFile(kubeadmconstants.GetKubeletKubeConfigPath())
-	if err != nil {
-		return err
+		// Now the kubelet will perform the TLS Bootstrap, transforming /etc/kubernetes/bootstrap-kubelet.conf to /etc/kubernetes/kubelet.conf
+		// Wait for the kubelet to create the /etc/kubernetes/kubelet.conf kubeconfig file. If this process
+		// times out, display a somewhat user-friendly message.
+		waiter := apiclient.NewKubeWaiter(nil, kubeadmconstants.TLSBootstrapTimeout, os.Stdout)
+		if err := waiter.WaitForKubeletAndFunc(waitForTLSBootstrappedClient); err != nil {
+			fmt.Printf(kubeadmJoinFailMsg, err)
+			return err
+		}
+
+		// When we know the /etc/kubernetes/kubelet.conf file is available, get the client
+		client, err := kubeconfigutil.ClientSetFromFile(kubeadmconstants.GetKubeletKubeConfigPath())
+		if err != nil {
+			return err
+		}
+
+		klog.V(1).Infoln("[kubelet-start] preserving the crisocket information for the node")
+		if err := patchnodephase.AnnotateCRISocket(client, cfg.NodeRegistration.Name, cfg.NodeRegistration.CRISocket); err != nil {
+			return errors.Wrap(err, "error uploading crisocket")
+		}
 	}
```

Comment:
for dryrun==true, we should mention what would happen.

```
fmt.Println("[dryrun] Would start the kubelet")
```

Reply:
yes, I should print more info, done